modules/extension/xsd/xsd-gml2/src/main/java/org/geotools/gml2/FeatureTypeCache.java

```diff
@@ -29,28 +29,19 @@
  * @source $URL$
  */
 public class FeatureTypeCache {
-    HashMap<Name, FeatureType> map = new HashMap<Name, FeatureType>();
+	ConcurrentHashMap<Name, FeatureType> map = new ConcurrentHashMap<Name, FeatureType>();
 
     public FeatureType get(Name name) {
-        synchronized (this) {
-            return map.get(name);
-        }
+    	return map.get(name);
     }
 
     public void put(FeatureType type) {
-        synchronized (this) {
-            if (map.get(type.getName()) != null) {
-                FeatureType other = map.get(type.getName());
-
-                if (!other.equals(type)) {
-                    String msg = "Type with same name already exists in cache.";
-                    throw new IllegalArgumentException(msg);
-                }
-
-                return;
-            }
-
-            map.put(type.getName(), type);
-        }
+    	FeatureType other = map.putIfAbsent(type.getName(), type);
+    	if(other != null) {
+    		if(!other.equals(type)) {
```

Comment:
Could you fix the extra indentation here? Does not adhere to the coding conventions.
Also, please squash all the commits into one (you can do a push --force to update your pull request after the squash)